proxy-server/go/proxy-main.go

```diff
@@ -0,0 +1,306 @@
+	})
+}
+
+// The default fallback handler.
+func handleRequestFallBack(res http.ResponseWriter, req *http.Request) {
+	res.WriteHeader(http.StatusUnauthorized)
+	//		res.WriteHeader(http.StatusBadRequest)
+	fmt.Fprintf(res, "\nWe are sorry, access denied.")
+	return
+}
+
+func validateAuthToken(token *extendedClaims) bool {
+	// token extra validation. its signature is already verified at this point
+	// considerations: -
+	//			check expiration time for this token
+	//          any extra user access limitations
+	return true
+}
+
+func redirectToFallBack(res http.ResponseWriter, req *http.Request, error int, origURL string) {
+
+	if debug {
+		log.Printf("fallback code: %d", error)
+	}
+	switch error {
+	case errNoCookie:
+		log.Printf("Fallback: no cookie token presented or bad cookie")
+	case errExpired:
+		log.Printf("Fallback: cookie expired")
+	case errSignatureInvalid:
+		log.Printf("Fallback: Unauthorized attempt to access resource: wrong signature\n")
+	case errTokenInvalid:
+		log.Printf("Fallback: Unauthorized attempt to access resource: invalid token\n")
+	case errGeneric:
+		log.Printf("Fallback: bad request")
+	}
+
+	uniqToken, _ := generateCookieSigningKey(40)
+
+	mutex.Lock()
+	tokenMap[string(uniqToken)] = ""
+	log.Printf("Added token %s", uniqToken)
+	mutex.Unlock()
+
+	u := fallbackURL + "?orig_request=" + origURL + "&ret_token=" + string(uniqToken)
+
+	if debug {
+		log.Printf("Redirecting to fallback url: %s", u)
+	}
+
+	http.Redirect(res, req, u, 301)
+}
+
+// Given a request send it to the appropriate url
+func handleRequestAndRedirect(res http.ResponseWriter, req *http.Request) {
+
+	targetURL := defaultURL
+
+	// if we receive authorization token in the request : the user authorized by the backend
+	satToken := req.URL.Query().Get("saturn_token")
+
+	if satToken != "" {
+		claims := &extendedClaims{}
+		tkn, err := jwt.ParseWithClaims(satToken, claims, func(token *jwt.Token) (interface{}, error) {
+			return sharedKey, nil
+		})
+
+		if err != nil || !tkn.Valid {
+			log.Printf("Authorization Error: invalid token from saturn")
+			// Breaking the loop. No more redirects. All is bad for this request.
+			res.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+
+		if len(claims.resourceAddress) != 0 {
+			targetURL = claims.resourceAddress
+		}
+
+		if len(claims.Issuer) == 0 {
+			log.Printf("Authorization Error: invalid token from saturn")
+			res.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+
+		mutex.Lock()
+		_, prs := tokenMap[claims.Issuer]
+		if prs {
+			delete(tokenMap, claims.Issuer)
+			log.Printf("Deleted token %s", claims.Issuer)
+
+		}
+		mutex.Unlock()
+
+		if !prs {
+			log.Printf("Authorization Error: invalid token from saturn")
+			res.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+
+		if validateAuthToken(claims) {
+			setNewCookie(res, req)
+			log.Printf("OK: Setting cookie")
+			serveReverseProxy(targetURL, res, req)
+			log.Printf("OK: Proxying to url: %s", targetURL)
+		}
```

Comment:
should have a `return` after this

Reply:
>should have a return after this
naked return? yes, but  there is no return value for this function..